### PR TITLE
Reject process instance migration if any element is mapped to a different type

### LINKED ISSUE #15125: Reject process instance migration if any element is mapped to a different type
In the `ProcessInstanceMigration` command processor, reject the command if:
- any of the mapping instructions of the command refer to a source and a target element with different element type, or different event type

Rejecting a command includes both appending a rejection with the rejection writer, as well as responding the rejection with the response writer.

Some tests must showcase this behavior.

Blocked by:
- https://github.com/camunda/zeebe/issues/15111

### PR DESCRIPTION
**Pull request by berkaycanbc**
## Description

Retrieving `targetElementType` from process definition can be re-used from PR https://github.com/camunda/zeebe/pull/15500. This PR can still be reviewed separately, and we can merge them later on.

## Related issues

closes #15125 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/MigrateProcessInstanceRejectionTest.java [hunk 1]
```diff
@@ -146,6 +146,58 @@ public void shouldRejectCommandWhenActiveElementIsNotMapped() {
...
+    assertThat(rejectionRecord)
+        .hasIntent(ProcessInstanceMigrationIntent.MIGRATE)
+        .hasRejectionType(RejectionType.INVALID_STATE)
+        .hasRejectionReason(
+            String.format(
+                """
+              Expected to migrate process instance '%s' \
+              but active element with id 'A' and type 'SERVICE_TASK' is mapped to \
+              an element with id 'A' and different type 'USER_TASK'. \
+              Active elements must be mapped to the same type.""",
```
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/MigrateProcessInstanceRejectionTest.java:196]: >Active elements must be mapped to the same type.

🔧 Elements must be mapped to elements of the same type.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java [hunk 2]
```diff
@@ -249,4 +265,31 @@ private static final class UnmappedActiveElementException extends RuntimeExcepti
               processInstanceKey, elementId));
     }
   }
+
+  /**
+   * Exception that can be thrown during the migration of a process instance, in case any of the
+   * mapping instructions of the command refer to a source and a target element with different
+   * element type, or different event type.
+   */
+  private static final class IncorrectMappingException extends RuntimeException {
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java:274]: 🔧 We can use a more expressive exception name, for example, `ElementTypeChangedException` or so.

**Review by korthout — APPROVED** [review]
Nice and clean set of changes! 🏅 

🔧 I only have a few non-blocking suggestions. Please have a look before merging.

LGTM 👍

**Comment by berkaycanbc** [comment]
bors merge

**Comment by berkaycanbc** [comment]
@korthout I somehow skip to apply suggested changes for this PR. I will send them along with #15461. 👍

**Pull request merged**